nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py

```diff
@@ -176,3 +177,108 @@ def _group_metric_calculate(self, group_metrics: Union[Dict[str, Tensor], List[T
         return group_sum_metric
+
+
+class AttentionSparsityAllocator(SparsityAllocator):
+    """
+    A specific allocator for attention has a structure like self-attention in Bert.
+    In this sparsity allocator, :math:`Q_{proj}`, :math:`K_{proj}`, :math:`V_{proj}` of the same attention head will in one group.
+    Sparsity allocator will combine their metrics and share the mask among these three weights in one group.
+    Note that by default, assuming QKV are implemented with `torch.nn.Linear`.
+    """
+    def __init__(self, pruner: Pruner, dim: int, dummy_input: Any):
+        super().__init__(pruner, dim=dim)
+        self.dummy_input = dummy_input
+
+    def _get_dependency(self):
+        """
+        Populate self.attention_name_groups by running inference on the module graph.
+        Currently, the group inferred AttentionWeightDependency is limited to a set of four weights, with the first
+        three corresponding to :math:`Q_{proj}`, :math:`K_{proj}`, :math:`V_{proj}` (in any order) and the last one being :math:`{Output}_{proj}`.
+        """
+        try:
+            graph = self.pruner.generate_graph(dummy_input=self.dummy_input)
+            dependency_tracer = AttentionWeightDependency(traced_model=graph.trace)
+            self.attention_name_groups = dependency_tracer.dependency_sets
+
+        except Exception as e:
+            raise RuntimeError('Graph trace failed: please check dummy_input, or specify attention_name_groups.\n'
+                               'Exception message: ' + str(e))
+
+    def _group_metric_calculate(self, group_metrics_dict: Dict[str, Tensor]) -> Tensor:
+        """
+        Sum :math:`Q_{proj}`, :math:`K_{proj}`, :math:`V_{proj}` metric value in the same position in one group.
+        """
+        group_metrics_list = list(group_metrics_dict.values())
+        assert all(group_metrics_list[0].size() == group_metric.size() for group_metric in group_metrics_list), 'Metrics size mismatched.'
+        group_sum_metric = torch.zeros(group_metrics_list[0].size(), device=group_metrics_list[0].device)
+        for group_metric in group_metrics_list:
+            group_sum_metric += group_metric
+        return group_sum_metric
+
+    def generate_sparsity(self, metrics: Dict) -> Dict[str, Dict[str, Tensor]]:
+        self._get_dependency()
+        masks = {}
+        grouped_metrics = {}
+
+        # Group the QKV metrics in one head.
+        for idx, names in enumerate(self.attention_name_groups):
+            group_metric_dict = OrderedDict()
+            for name in names[0: 3]:
+                # If the Linear will be pruned, add it to this group.
+                if name in metrics:
+                    group_metric_dict[name] = metrics[name]
+                    if self.continuous_mask:
+                        group_metric_dict[name] *= self._compress_mask(self.pruner.get_modules_wrapper()[name].weight_mask)
+            grouped_metrics[idx] = group_metric_dict
```

Comment:
It seems that we don't support `op_names` in attention mode. If module is pruned by setting `op_names` in config_list while `self.attention_name_groups` recording all name of Q, K, V, error may be raised on line 237 since some value of `grouped_metrics` will be empty.

Reply:
you are right, this place will raise an error, I will fix it.